sound/soc/sof/ipc4-control.c

```diff
@@ -181,32 +181,262 @@ static int sof_ipc4_volume_get(struct snd_sof_control *scontrol,
 	return 0;
 }
 
+static int sof_ipc4_set_get_bytes_data(struct snd_sof_dev *sdev,
+				       struct snd_sof_control *scontrol,
+				       bool set, bool lock)
+{
+	struct sof_ipc4_control_data *cdata = scontrol->ipc_control_data;
+	struct sof_abi_hdr *data = cdata->data;
+	struct sof_ipc4_msg *msg = &cdata->msg;
+	int ret = 0;
+
+	/* Send the new data to the firmware only if it is powered up */
+	if (set && !pm_runtime_active(sdev->dev))
+		return 0;
+
+	msg->extension = SOF_IPC4_MOD_EXT_MSG_PARAM_ID(data->ipc4.param_id);
+
+	msg->data_ptr = data->data;
+	msg->data_size = data->size;
```

Comment:
Hrm, interesting. We should be sending the sof_abi_hdr part as well to the firmware, not only the blob?

In IPC3 we have `struct sof_ipc_ctrl_data` which wraps all sorts of kcontrol related data (volume, enum, bytes), at the end of `struct sof_ipc_ctrl_data` we have a union and one option is the `sof_abi_hdr` (sof_ipc_ctrl_value_chan / sof_ipc_ctrl_value_comp / sof_abi_hdr)

With IPC4 we don't have such a generic kcontrol message, but we send everything with large_config and the target  module and it's parameter will tell the recipient what the data is and it will know how to handle it.

It is simple to send the header along, but it will need firmware side changes to handle the additional wrap (and btw, the abi_hdr use in firmware is just to `data->data` to get to the real blob and the header's content ignored.

Reply:
Right, after a quick check in my brain: it does not make sense to send the sof_abi_hdr along with the blobs. It could be only understood by SOF modules and IPC4 is much broader, we can not start sending blobs with strange headers to existing modules.